presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/OptimizeMixedDistinctAggregations.java

```diff
@@ -0,0 +1,507 @@
+import com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.GroupIdNode;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.ComparisonExpressionType;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.IfExpression;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NullLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+
+/*
+ * This optimizer convert query of form:
+ *
+ *  SELECT a1, a2,..., an, F1(b1), F2(b2), F3(b3), ...., Fm(bm), F(distinct c) FROM Table GROUP BY a1, a2, ..., an
+ *
+ *  INTO
+ *
+ *  SELECT a1, a2,..., an, arbitrary(if(group_id = 0, f1)),...., arbitrary(if(group_id = 0, fm)), F(if(group_id = 1, c)) FROM
+ *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group_id FROM
+ *        SELECT a1, a2,..., an, b1, b2, ... ,bn, c FROM Table GROUP BY GROUPING SETS ((a1, a2,..., an, b1, b2, ... ,bn), (a1, a2,..., an, c))
+ *      GROUP BY a1, a2,..., an, c, group_id
+ *  GROUP BY a1, a2,..., an
+ */
+public class OptimizeMixedDistinctAggregations
+        implements Rule
+{
+    private final FunctionRegistry functionRegistry;
+
+    public OptimizeMixedDistinctAggregations(FunctionRegistry functionRegistry)
+    {
+        this.functionRegistry = functionRegistry;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+
+        AggregationNode aggregationNode = (AggregationNode) node;
+
+        // optimize if and only if
+        // some aggregation functions have a distinct mask symbol
+        // and if not all aggregation functions on same distinct mask symbol (this case handled by SingleMarkDistinctToGroupBy)
+        Set<Symbol> masks = ImmutableSet.copyOf(aggregationNode.getMasks().values());
+        if (masks.size() != 1 || aggregationNode.getMasks().size() == aggregationNode.getAggregations().size()) {
```

Comment:
updated the comment. The check means there should be only one distinct aggregation and there must be at least one non-distinct aggregation along with it.